core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/sync/MixedPackageVersionService.java

```diff
@@ -0,0 +1,226 @@
+
+    @Inject
+    private MixedPackageMessageProvider mixedPackageMessageProvider;
+
+    @Inject
+    private MixedPackageVersionComparator mixedPackageVersionComparator;
+
+    public void validatePackageVersions(Long stackId, CmSyncOperationResult syncResult, Set<Image> candidateImages) {
+        Optional<String> activeCmVersion = syncResult.getCmRepoSyncOperationResult().getInstalledCmVersion();
+        if (activeCmVersion.isPresent()) {
+            Set<ParcelInfo> activeParcels = syncResult.getCmParcelSyncOperationResult().getActiveParcels();
+            Optional<StatedImage> currentImage = findCurrentImage(stackId);
+            if (currentImage.isPresent()) {
+                if (areComponentVersionsAreEqualWithTheOriginalImage(currentImage.get(), activeCmVersion.get(), activeParcels)) {
+                    LOGGER.debug("The current component versions {} are the same as the versions on the cluster's image {}", activeParcels, currentImage);
+                } else {
+                    findTargetImage(stackId).ifPresentOrElse(
+                            image -> examinePackageVersionsWithTargetImage(stackId, image, activeCmVersion.get(), activeParcels),
+                            () -> examinePackageVersionsWithAllCandidateImages(stackId, candidateImages, activeCmVersion.get(), activeParcels,
+                                    getImageCatalogUrl(currentImage)));
+                }
+            } else {
+                examinePackageVersionsWithAllCandidateImages(stackId, candidateImages, activeCmVersion.get(), activeParcels, getImageCatalogUrl(currentImage));
+            }
+        } else {
+            LOGGER.warn("The validation of the package versions cannot be performed because the CM version is not present in the sync result. {}", syncResult);
+        }
+    }
+
+    private Optional<Image> findTargetImage(Long stackId) {
+        Optional<com.sequenceiq.cloudbreak.cloud.model.Image> targetImage = clusterUpgradeTargetImageService.findTargetImage(stackId);
+        return targetImage.isPresent() ? findImageInCatalog(targetImage.get()) : Optional.empty();
+    }
+
+    private Optional<Image> findImageInCatalog(com.sequenceiq.cloudbreak.cloud.model.Image image) {
+        try {
+            return Optional.of(imageCatalogService.getImage(image.getImageCatalogUrl(), image.getImageCatalogName(), image.getImageId()).getImage());
+        } catch (CloudbreakImageNotFoundException | CloudbreakImageCatalogException e) {
+            LOGGER.warn("There is no image {} found in catalog {}.", image.getImageId(), image.getImageCatalogUrl());
+            return Optional.empty();
+        }
+    }
+
+    private void examinePackageVersionsWithTargetImage(Long stackId, Image targetImage, String activeCmVersion, Set<ParcelInfo> activeParcels) {
+        LOGGER.debug("Comparing active package versions {} and active CM version {} with target image {}", activeParcels, activeCmVersion, targetImage);
```

Comment:
This is also a largish private method, and would add test for it. However, in that case it might make sense to put it to a separate class and add tests for it on its own.